pkg/operator/certrotation/signer_test.go

```diff
@@ -1,24 +1,219 @@
 package certrotation
 
 import (
+	"bytes"
 	"context"
+	"reflect"
 	"strings"
+	"sync"
 	"testing"
 	"time"
 
 	"github.com/davecgh/go-spew/spew"
 
 	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
 	kubefake "k8s.io/client-go/kubernetes/fake"
 	corev1listers "k8s.io/client-go/listers/core/v1"
 	clienttesting "k8s.io/client-go/testing"
 	"k8s.io/client-go/tools/cache"
 
 	"github.com/openshift/api/annotations"
+	"github.com/openshift/library-go/pkg/crypto"
 	"github.com/openshift/library-go/pkg/operator/events"
 )
 
+// loadbalancer-serving-signer secret in openshift-kube-apiserver-operator namespace
+// is reconciled by multiple controllers at the same time without any coordination.
+// it looks like we have `2` separate processes and `4` different controllers competing over the same resource.
+//
+// the following unit test demonstrates that the secret crypto material
+// can be regenerated, which has serious consequences for the platform
+// as it can break external clients and the cluster itself.
+//
+// controllers reconciling the secret:
+//  - https://github.com/openshift/cluster-kube-apiserver-operator/blob/release-4.15/pkg/operator/certrotationcontroller/certrotationcontroller.go#L338
+//  - https://github.com/openshift/cluster-kube-apiserver-operator/blob/release-4.15/pkg/operator/certrotationcontroller/certrotationcontroller.go#L392
+
+// processes reconciling the secret:
+//   - https://github.com/openshift/cluster-kube-apiserver-operator/blob/63691a0ee938b1b7002777db8464d50791efb94a/pkg/operator/starter.go#L318
+//   - https://github.com/openshift/cluster-kube-apiserver-operator/blob/052beea48fdeb9be7cdafe7a5ae2e7c4c74c3a0f/pkg/cmd/certregenerationcontroller/cmd.go#L105
+//
+// In general, NewCertRotationController react to changes on secrets in
+// openshift-kube-apiserver-operator ns (https://github.com/openshift/cluster-kube-apiserver-operator/blob/ae0c9631284bca8cfcbc305674781e097f7238ae/vendor/github.com/openshift/library-go/pkg/operator/certrotation/client_cert_rotation_controller.go#L94)
+func TestEnsureSigningCertKeyPairRace(t *testing.T) {
+	secondControllerStartChan := make(chan struct{})
+	secondControllerEndChan := make(chan struct{})
+	firstControllerEndChan := make(chan struct{})
+
+	// represents a secret that was created before 4.7 and
+	// hasn't been updated until now (upgrade to 4.15)
```

Comment:
Yes, the secret must not have `kubernetes.io/tls` because the` ensureSecretTLSTypeSet` method will prevent the delete and re-create path. However, the method is quite new and was added a few weeks ago. So, not all versions have this patch. Besides, clusters created **before** version `4.7` have a different secret type set.